internal/venice-consumer/src/main/java/com/linkedin/venice/client/consumer/VeniceAfterImageConsumerImpl.java

```diff
@@ -0,0 +1,176 @@
+      ChangelogClientConfig changelogClientConfig,
+      Consumer<KafkaKey, KafkaMessageEnvelope> kafkaConsumer) {
+    super(changelogClientConfig, kafkaConsumer);
+    this.currentTopic = Version.composeKafkaTopic(storeName, storeCurrentVersion);
+  }
+
+  public VeniceAfterImageConsumerImpl(ChangelogClientConfig changelogClientConfig) {
+    this(changelogClientConfig, new KafkaConsumer<>(changelogClientConfig.getConsumerProperties()));
+  }
+
+  @Override
+  public Collection<PubSubMessage<K, ChangeEvent<V>, Long>> poll(long timeoutInMs) {
+    List<PubSubMessage<K, ChangeEvent<V>, Long>> pubSubMessages = new ArrayList<>();
+    ConsumerRecords<KafkaKey, KafkaMessageEnvelope> consumerRecords = kafkaConsumer.poll(timeoutInMs);
+    for (ConsumerRecord<KafkaKey, KafkaMessageEnvelope> consumerRecord: consumerRecords) {
+      PubSubTopicPartition pubSubTopicPartition = getPubSubTopicPartitionFromConsumerRecord(consumerRecord);
+      long offset = consumerRecord.offset();
+      int payloadSize = consumerRecord.serializedKeySize() + consumerRecord.serializedValueSize();
+      long timestamp = consumerRecord.timestamp();
+      if (consumerRecord.key().isControlMessage()) {
+        ControlMessage controlMessage = (ControlMessage) consumerRecord.value().payloadUnion;
+        if (handleControlMessage(controlMessage, pubSubTopicPartition)) {
+          return pubSubMessages;
+        }
+      } else {
+        byte[] keyBytes = consumerRecord.key().getKey();
+        K currentKey = keyDeserializer.deserialize(keyBytes);
+        if (isReadFromChangeCaptureTopic) {
+          Optional<PubSubMessage<K, ChangeEvent<V>, Long>> pubSubMessage =
+              convertConsumerRecordToPubSubChangeEventMessage(consumerRecord, pubSubTopicPartition);
+          if (pubSubMessage.isPresent()) {
+            pubSubMessages.add(
+                new ImmutablePubSubMessage<>(
+                    pubSubMessage.get().getKey(),
+                    new ChangeEvent<>(null, pubSubMessage.get().getValue().getCurrentValue()),
+                    pubSubTopicPartition,
+                    offset,
+                    timestamp,
+                    keyBytes.length + currentValuePayloadSize[pubSubTopicPartition.getPartitionNumber()]));
+          }
+        } else {
+          PubSubMessage<K, ChangeEvent<V>, Long> pubSubMessage = convertRecordToPubSubMessage(
+              consumerRecord.value(),
+              currentKey,
+              pubSubTopicPartition,
+              offset,
+              timestamp,
+              payloadSize);
+          pubSubMessages.add(pubSubMessage);
+        }
+      }
+    }
+    return pubSubMessages;
+  }
+
+  private boolean handleControlMessage(ControlMessage controlMessage, PubSubTopicPartition pubSubTopicPartition) {
+    ControlMessageType controlMessageType = ControlMessageType.valueOf(controlMessage);
+    if (controlMessageType.equals(ControlMessageType.END_OF_PUSH)) {
+      isReadFromChangeCaptureTopic = true;
```

Comment:
This is a touch confusing I think.  It looks like what we'll end up doing is consuming on the version topic, then when we hit end of push, we'll jump to the change capture topic.  We express that state with this member variable.  However, we use this member variable to determine the type of PubSubMessage we return (convertRecordToPubSubMessage() vs. convertConsumerRecordToPubSubChangeEventMessage).  If this is a user which is leveraging before/after image, they would expect to get the pubSubChangeEventMEssage every time regardless of where we're consuming from (in the case where we consume from VT the before image is just null).

Reply:
The flag of `isReadFromChangeCaptureTopic` is not for reading from which topics.  I just change the interface of `VeniceChangelogConsumer#poll` to return the events as `PubSubMessage<K, ChangeEvent<V>, Long>` instead of `PubSubMessage`. The reason I change is to avoid too much cast operations on PubSubMessage's values. Now when a user decide to consume after image, they should expect the before image is null.